api/src/main/java/org/pmiops/workbench/api/ProfileController.java

```diff
@@ -481,6 +488,49 @@ private User initializeUserIfNeeded() {
     return getProfileResponse(saveUserWithConflictHandling(user));
   }
 
+  /**
+   * This methods updates logged in user's training status from Moodle.
+   * 1. Check if user have moodle_id,
+   *    a. if not retrieve it from MOODLE API and save it in the Database
+   * 2. Using the MOODLE_ID get all of the user's Badge
+   * 3. If user has earned All of Us badge, update the database with
+   *    a. training completion time as current time
+   *    b. training expiration date with as returned from MOODLE.
+   * @return Profile updated with training completion time
+   */
+  @Override
+  public ResponseEntity<Profile> syncTrainingStatus() {
+    User user = userProvider.get();
+    Profile profile = profileService.getProfile(user);
+    try {
+      Integer moodleId = user.getMoodleId();
+      if (moodleId == null) {
+        moodleId = complianceService.getMoodleId(user.getEmail());
+        user.setMoodleId(moodleId);
+      }
+      List<BadgeDetails> badgeResponse = complianceService.getUserBadge(moodleId);
+      if (badgeResponse != null && badgeResponse.size() > 0) {
+        BadgeDetails badge = badgeResponse.get(0);
+        if (badge.getDateexpire() == null) {
+          //This can happen if date expire is set to never
+          user.setTrainingExpirationTime(null);
+        } else {
+          user.setTrainingExpirationTime(new Timestamp(Long.parseLong(badge.getDateexpire())));
+        }
+        Timestamp now = new Timestamp(clock.instant().toEpochMilli());
+        user.setTrainingCompletionTime(now);
+        profile.setTrainingCompletionTime(now.getTime());
+      }
+      userDao.save(user);
+    } catch (ApiException ex) {
+      if (ex.getCode() == HttpStatus.NOT_FOUND.value()) {
+        throw new NotFoundException(ex.getMessage());
```

Comment:
For RW users who haven't yet logged into Moodle, we'll be returning a 404 error here. I'm not sure that's the right thing to do, since in effect the training status sync is a no-op at that point: a user who doesn't have a Moodle user account yet has by definition not completed their training.

I think we have a couple options:

1) Have a separate try/catch around lines 506-510, to catch a NOT_FOUND response from the Moodle API. In that case, return a normal ResponseEntity. And continue throwing a NotFoundException for other Moodle API calls, since that's an unexpected error.

2) Keep as-is and have callers of this method handle the NotFoundException.

My gut feeling says #1 is the way to go, but I don't have a super strong preference.

Reply:
Good point. I missed this scenario and i agree point 1 is the right way to go.

Reply:
After thinking about it, i think it would be better to throw Not Found Exception, so UI can show appropriate message (create user in Moodle ) etc